api/src/services/report/validate.js

```diff
@@ -1,67 +1,37 @@
-const _ = require('underscore');
-
-/*
- *  return array of errors or empty array.
- */
-exports.validate = (def, form_data) => {
-  const missing_fields = [];
-
-  const _isDefined = obj => !_.isUndefined(obj) && !_.isNull(obj);
-
-  _.each(def.fields, (field, k) => {
-    const orig_key = k;
-    let key = k.split('.');
-    let data = form_data;
-
-    while(key.length > 1) {
-      data = data[key.shift()];
-    }
-
-    key = key[0];
+const exists = function(value) {
+  return value !== undefined && value !== null;
+};
 
-    if (field.required) {
-      if (
-        !data ||
-        !_isDefined(data[key]) ||
-        (_.isArray(data[key]) && !_isDefined(data[key][0]))
-      ) {
-        missing_fields.push(orig_key);
+exports.validate = (def, formData) => {
+  const errors = [];
+  const missingFields = [];
+
+  Object.keys(def.fields).forEach(k => {
+    const field = def.fields[k];
+    const value = formData[k];
+
+    if (field.required && !exists(value)) {
+      missingFields.push(k);
+    } else if (exists(value) && !field.list) {
+      // list is an undocumented but currently used feature of SMS parsing, where it converts
+      // integers into a translated string. Validation runs after parsing for SMS, so a field with a
+      // list parameter would fail validation.
+
+      if (field.type === 'integer' && (typeof value !== 'number' || !Number.isInteger(value))) {
```

Comment:
previously we never actually checked field types for JSON-backed JSON reports, which seems a little lazy. I'm not married to doing it though.